db-engine/src/main/java/com/splicemachine/db/iapi/sql/dictionary/foreignkey/Graph.java

```diff
@@ -208,6 +208,7 @@ public void removeEdge(int from, int to) {
                     edgeNode.next = null; // bye, GC.
                     return;
                 }
+                previous = edgeNode;
```

Comment:
Is this a legitimate issue on the current version? What kind of failures would it cause if so?

Reply:
Yes it is, it causes the graph to miss some edges (which leads to incorrect conclusions about the validity of the graph). Since we're not updating `previous` correctly, we end up removing the wrong edge (line 207). Here is an example that exposes this bug:

```
digraph {
  C -> A[label="Restrict"];
  A -> B[label="Cascade"];
  C -> A[label="SetNull"];
  B -> C[label="Cascade"];
}
```

In this graph, adding the final edge `B -> C` causes two cycles to appear:
1. `A -Cascade-> B -Cascade-> C -SetNull-> A`
2. `A -Cascade-> B -Cascade-> C -Restrict-> A`

We remove the first cycle by rerouting `C-SetNull->A` to surrogate node `C-SetNull->S_A_0`. At this point, the edge list for node `C` looks something like this:
`'C' -> 'S_A_0 (setnull)' -> 'A (restrict)'`.

Now, we try check again if we have a cycle (line 124), and we find that we have another one: `A -Cascade-> B -Cascade-> C -Restrict-> A`. So now we reroute `C -Restrict-> A` to a new surrogate node `C -Restrict->S_A_1`. (at this point, the edge list for node `C` will be `'C' -> 'S_A_1 (restrict)' -> 'S_A_0 (setnull)' -> 'A (restrict)'`) And we attempt to remove the edge between `C` and `A` (line 122).

Here is where we hit the problem, when `removeEdge` is called we set the `previous` to edge `'S_A_1 (restrict)'`, and we don't update it anymore, so when we find the edge to remove, we end removing this edge instead of the edge `'A (restrict)'`, therefore we end up building a different graph which happens to be OK.

Reply:
Failures it causes is accepting invalid graphs, I think this is unlikely to happen with customers who are migrating (hopefully) valid schemas from other DBMS to SpliceMachine.